Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30502 [BC-BREAKING] change index_select scalar_check to retain dimensionality of input.
* **#30438 Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM.**
* #30437 [WIP] Remove scalar_check from topk, move it to the THC implementation.
* #30436 Turn off scalar_checks for _th_reciprocal.
* #30435 Turn off scalar_checks for torch.clamp.
* #30434 Turn off scalar_checks for exp, cos, cosh, tan, atan, tanh, erf, erfc.

The input(s) can't be 0-dimensional, so its irrelevant.

Differential Revision: [D18699090](https://our.internmc.facebook.com/intern/diff/D18699090)

gchanan committed:
Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM.

The input(s) can't be 0-dimensional, so its irrelevant.

[ghstack-poisoned]

gchanan committed:
Update on "Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM."


The input(s) can't be 0-dimensional, so its irrelevant.

[ghstack-poisoned]

gchanan committed:
Update on "Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM."


The input(s) can't be 0-dimensional, so its irrelevant.

[ghstack-poisoned]

gchanan committed:
Update on "Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM."


The input(s) can't be 0-dimensional, so its irrelevant.

Differential Revision: [D18699090](https://our.internmc.facebook.com/intern/diff/D18699090)

[ghstack-poisoned]

gchanan committed:
Update on "Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM."


The input(s) can't be 0-dimensional, so its irrelevant.

Differential Revision: [D18699090](https://our.internmc.facebook.com/intern/diff/D18699090)

[ghstack-poisoned]

albanD commented on test/test_torch.py:
```diff
@@ -6258,6 +6258,12 @@ def test_scalar_check(self, device):
+        # convolutions.  Yes, we are testing nn.functional here; seems justified
+        # given its similar to the other tests
+        w = torch.randn(2, 1, 3, 3, device=device).div_(2).requires_grad_()
+        self.assertRaises(RuntimeError, lambda: torch.nn.functional.conv2d(zero_d, w, groups=1))
```
This will always be routed to the `thnn` version? What about the mklnn if it's available?

gchanan replied:
MKLDNN is only on CPU.

For CUDA it depends on cudnn is available or not -- I'm not sure if we have any tests of that case but it doesn't _really_ matter -- what matters is that the function handles scalars correctly.

albanD replied:
Is this test CPU or GPU?

In any case, since we have different backends that all need to handle this case properly, we should test all the backends no?

gchanan replied:
I don't know what backend means, but this test does check all the device types.

albanD replied:
My understanding is that now that the codegen does not take care of this but the low level implementation does. Every single possible low level implementation must handle this properly.

As an example:
If we run this on GPU. either cudnn or thcunn will be checked. But not both. Right?

gchanan committed:
Update on "Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM."


The input(s) can't be 0-dimensional, so its irrelevant.

Differential Revision: [D18699090](https://our.internmc.facebook.com/intern/diff/D18699090)

[ghstack-poisoned]

gchanan committed:
Update on "Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM."


The input(s) can't be 0-dimensional, so its irrelevant.

Differential Revision: [D18699090](https://our.internmc.facebook.com/intern/diff/D18699090)

[ghstack-poisoned]

gchanan committed:
Update on "Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM."


The input(s) can't be 0-dimensional, so its irrelevant.

Differential Revision: [D18699090](https://our.internmc.facebook.com/intern/diff/D18699090)

[ghstack-poisoned]

gchanan committed:
Update on "Turn off scalar_checks for SpatialDepthwiseConvolution and SpatialConvolutionMM."


The input(s) can't be 0-dimensional, so its irrelevant.

Differential Revision: [D18699090](https://our.internmc.facebook.com/intern/diff/D18699090)

[ghstack-poisoned]
